Database/MySQL/PreparedStatement.hs

```diff
@@ -0,0 +1,485 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+module Database.MySQL.PreparedStatement
+    where
+
+import           Control.Monad             (void, when)
+import           Data.ByteString.Char8
+import           Data.ByteString.Internal  (create)
+import           Data.ByteString.Unsafe    (unsafeUseAsCStringLen)
+import qualified Data.Fixed                as Fixed
+import           Data.Text                 (Text)
+import qualified Data.Text.Foreign         as Text
+import           Data.Time.Calendar        (Day, fromGregorian, toGregorian)
+import           Data.Time.LocalTime       (LocalTime (..), TimeOfDay (..),
+                                            makeTimeOfDayValid)
+import           Database.MySQL.Base.C
+import           Database.MySQL.Base.Types hiding (Type (..))
+import qualified Database.MySQL.Base.Types as Types
+import           Foreign                   hiding (newForeignPtr, void)
+import           Foreign.C.Error
+import           Foreign.C.String          (peekCString)
+import           Foreign.Concurrent        (newForeignPtr)
+
+data Statement = Statement
+    { connectionPtr :: ForeignPtr MYSQL
+    , statementPtr  :: ForeignPtr MYSQL_STMT
+    }
+
+prepare :: ForeignPtr MYSQL -> Text -> IO Statement
+prepare mysql q =
+    withForeignPtr mysql $ \mysql' -> do
+    mysqlStmt <- throwIfNull "mysql_stmt_init" $ mysql_stmt_init mysql'
+    stmt <- newForeignPtr mysqlStmt (void $ mysql_stmt_close mysqlStmt)
+    Text.withCStringLen q $ \(p,l) -> do
+        res <- mysql_stmt_prepare mysqlStmt p (fromIntegral l)
+        when (res > 0) $ do
+            err <- peekCString =<< mysql_stmt_error mysqlStmt
+            error err
+    pure $ Statement mysql stmt
+
+withConn :: Statement -> (Ptr MYSQL -> IO a) -> IO a
+withConn statement =
+    withForeignPtr (connectionPtr statement)
+
+withStatement :: Statement -> (Ptr MYSQL_STMT -> IO a) -> IO a
+withStatement statement =
+    withForeignPtr (statementPtr statement)
+
+data Value
+    = Null
+    | TinyInt Int8
+    | Short Int16
+    | MediumInt Int32
+    | Long Int32
+    | LongLong Int64
+    | UnsignedTinyInt Word8
+    | UnsignedShort Word16
+    | UnsignedMediumInt Word32
+    | UnsignedLong Word32
+    | UnsignedLongLong Word64
+    | Float Float
+    | Double Double
+    | Decimal ByteString
+    | Time TimeOfDay
+    | Date Day
+    | DateTime LocalTime
+    | Timestamp LocalTime
+    | String Text
+    | Blob ByteString
+    deriving (Eq, Show)
+
+withValue :: Value -> (MYSQL_BIND -> IO a) -> IO a
```

Comment:
Is there a better way to implement this or is this kind of boilerplate inevitable?